Document Agent Skills feature in ADK Go (#1687)
google/adk-docs: docs/skills/index.md

impact on the operating context window of the agent.

!!! example "Experimental"
    The Skills feature is experimental. We welcome your feedback via the
    respective ADK GitHub repositories:
    [feedback for Python](https://github.com/google/adk-python/issues/new?template=feature_request.md&labels=skills),

>     [feedback for Go](https://github.com/google/adk-go/issues/new?template=feature_request.md&labels=skills)

Reviewer: ```suggestion
    The Skills feature is experimental. We welcome your feedback via the
    respective ADK GitHub repositories:
    [ADK Python](https://github.com/google/adk-python/issues/new?template=feature_request.md&labels=skills),
    [ADK TypeScript](https://github.com/google/adk-js/issues/new?template=feature_request.md&labels=skills),
    [ADK Go](https://github.com/google/adk-go/issues/new?template=feature_request.md&labels=skills).
```

Updating this note.

    )
    ```

    For a complete code example of an ADK agent with a Skill, including both
    file-based and in-line Skill definitions, see the code sample
    [skills_agent](https://github.com/google/adk-python/tree/main/contributing/samples/skills_agent).

=== "TypeScript"

    ```typescript
    --8<-- "examples/typescript/snippets/skills/get_started.ts:full_example"
    ```

=== "Go"

    ```go
    import (
        "context"
        "os"

>         "google.golang.org/adk/agent"

Reviewer: The "google.golang.org/adk/agent" import is unused in this snippet. Suggest that we remove it from this code sample to prevent compile errors.

        documentation, templates, or examples.
    -   `scripts/`: Executable scripts supported by the agent runtime.

> ### Recommended Skills directory structure

Reviewer: ```suggestion
### Skills directory structure
```

Suggest using a shorter sub-heading.

    --8<-- "examples/typescript/snippets/skills/inline_skill.ts:full_example"
    ```

=== "Go"

    !!! note
        ADK Go does not currently provide a standard Source for inline skills,
        though this may be added in the future.
        To define skills directly in code, you must implement the `skill.Source`
        interface yourself, as shown below.

>     ```go

Reviewer: This example is missing its import block. It uses `context`, `io`, `strings`, and `slices` but none are shown. The other Go snippets in this PR include imports, so this one should too for consistency.

        }
        return []string{"references/hello_world.txt", "references/example.md"}, nil
    }

    func (s *StaticSource) LoadResource(ctx context.Context, name, resourcePath string) (io.ReadCloser, error) {
        if name != "greeting-skill" {
            return nil, skill.ErrSkillNotFound
        }
        switch resourcePath {
        case "references/hello_world.txt":
            return io.NopCloser(strings.NewReader("Hello! So glad to have you here!")), nil
        case "references/example.md":
            return io.NopCloser(strings.NewReader("This is an example reference.")), nil
        default:
            return nil, skill.ErrResourceNotFound
        }
    }
    ```

### Read Skills from filesystem {#filesystem-skills}

=== "Python"

    ```python
    import pathlib

    from google.adk.skills import load_skill_from_dir
    from google.adk.tools import skill_toolset

    greeting_skill = load_skill_from_dir(
        pathlib.Path(__file__).parent / "skills" / "greeting-skill"
    )
    weather_skill = load_skill_from_dir(
        pathlib.Path(__file__).parent / "skills" / "weather-skill"
    )

    my_skill_toolset = skill_toolset.SkillToolset(

>         skills=[weather_skill, greating_skill],

Reviewer: ```suggestion
        skills=[weather_skill, greeting_skill],
```

    )
    ```

=== "Go"

    ```go
    import (
        "os"
    
        "google.golang.org/adk/tool/skilltoolset/skill"
        "google.golang.org/adk/tool/skilltoolset"
    )
    
    // ...

    source := skill.NewFileSystemSource(os.DirFS("./skills"))

    // This example doesn't use any optional wrappers, but you can use them if
    // needed, e.g.:

>     //   source, _, err = skill.WithFrontmatterPreloadSource(source)

Reviewer 2: Don't we miss the `ctx` parameter?

Author: Yes, we were. Fixed now

>     //   source, _, err = skill.WithCompletePreloadSource(source)

Reviewer: These commented-out examples seem to be missing the `ctx` parameter.

```suggestion
    //   source, _, err = skill.WithFrontmatterPreloadSource(ctx, source)
    //   source, _, err = skill.WithCompletePreloadSource(ctx, source)
```

    // For more information about these and other wrappers, see
    // https://pkg.go.dev/google.golang.org/adk/tool/skilltoolset/skill#Source.

    skillToolset, err := skilltoolset.New(ctx, skilltoolset.Config{
        Source: source,
    })
    if err != nil {
        // handle error
    }
    ```

> ### Advanced. Provision Skills dynamically {#dynamic-skills}

Reviewer: ```suggestion
### Provision Skills dynamically {#dynamic-skills}
```

Remove the "Advanced." portion from the section heading, it is not necessary and the formatting is odd.

When starting out, defining skills statically—whether through local files,
embedded assets, or in-line code definitions—is usually everything you need.

However, as you move to production, you might face two common challenges:

1.  **Enabling zero-downtime skill updates:** Rebuilding and deploying code
    every time a skill definition or reference asset needs a slight adjustment
    slows your team down.
2.  **Running A/B experiments and skill personalization:** Giving every user
    the exact same static skills makes it difficult to run targeted experiments
    or tailor the skill set per tenant.

=== "Go"

**Custom skill.Source implementation**

    -   **Live updates:** By writing a custom `Source` connected to a database
        or a custom deployment pipeline, your backend fetches the freshest skill
        components on demand, letting you update skills instantly without ever
        restarting your server.
    -   **Targeted skills:** Because your custom `Source` runs normal Go code
        for each incoming query, it can check the current context and decide
        exactly which skills to return, making basic A/B testing and

>         personalization easy to build.

Reviewer: The formatting under === "Go" is broken (the content is not indented for MkDocs tabs). You can view the docs preview build to verify.

> 

Reviewer: This entire section is difficult for readers to parse. It restates the two challenges from the intro and says a custom `skill.Source` solves them, but the reader already saw a full `skill.Source` implementation in the inline skills section above.

This section would be more useful if it either showed a concrete example (e.g., a source backed by a database) or was condensed into a brief note (not specific to a particular language) in the inline skills section pointing out that Source can be backed by any data store for dynamic use cases.

For the purposes of this PR, I would consider either removing this section, or changing to a note in the inline skills section.

## Next steps

Check out these resources for building agents with Skills:

*   [Skills in Python - code sample](https://github.com/google/adk-python/tree/main/contributing/samples/skills_agent).
*   [Skills in Go - code sample](https://github.com/google/adk-go/tree/main/examples/skills).

> *   Agent Skills [specification documentation](https://agentskills.io/).

Reviewer: ```suggestion
- [Skills in Python - code sample](https://github.com/google/adk-python/tree/main/contributing/samples/skills_agent)
- [Skills in Go - code sample](https://github.com/google/adk-go/tree/main/examples/skills)
- Agent Skills [specification documentation](https://agentskills.io/)
```
